### PR TITLE
PermsSyncer: track recent job states in redis

### PR DESCRIPTION
**Pull request by bobheadxi**
Tracks perms sync job outcomes, notably per-provider states (https://github.com/sourcegraph/sourcegraph/pull/44257), in Redis. We set an aggressive TTL of 5 minutes by default because in practice, only the details for the last few sync jobs are relevant, and this prevents us from putting too much pressure on Redis on larger instances.

This can be extended in the future to be database-backed, but for now I think Redis is the right choice - the volume of entries can be quite high when accumulated so we want something easy to expire, and realistically only the details for the last few sync jobs are relevant. Additionally, the records store and reader in package `synclogs` can easily be rewritten without too much impact to write to a database in the future.

The goal is to eventually allow us to render this data via the GraphQL API similarly to the external service sync status API  (https://github.com/sourcegraph/sourcegraph/pull/44387), hence the placement of the struct definitions in a shared package

Part of https://github.com/sourcegraph/customer/issues/1534
Stacked on https://github.com/sourcegraph/sourcegraph/pull/44257

## Test plan

Add a GitHub code host with `"authorization": {}`, and add a private repo to the list

Wait a few minutes

```sh
➜  redis-cli KEYS 'v2:authz/sync-job-records:*'
1) "v2:authz/sync-job-records:1668564699013272000"                 
v2:authz/sync-job-records:repo-8
➜  redis-cli GET v2:authz/sync-job-records:1668564699013272000 | jq
{
  "request_type": "repo",
  "request_id": 8,
  "completed": "2022-11-16T02:11:39.013272Z",
  "status": "SUCCESS",
  "message": "",
  "providers": [
    {
      "provider_id": "https://github.com/",
      "provider_type": "github",
      "status": "SUCCESS",
      "message": "FetchRepoPerms"
    }
  ]
}
```

### CODE REVIEW

**Comment by bobheadxi** [comment]
Current dependencies on/for this PR:
* main
  * **PR #44251** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/44251" target="_blank">Graphite</a> 
    * **PR #44257** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/44257" target="_blank">Graphite</a> 
      * **PR #44258** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/44258" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/44258?utm_source=stack-comment).

**Review by unknwon — COMMENTED** [review]
The approach does seem to make sense to me, will do a thorough review once marked as ready 👍

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer.go [hunk 1]
```diff
@@ -43,6 +44,8 @@ var scheduleReposCounter = promauto.NewCounter(prometheus.CounterOpts{
 	Help: "Counts number of repos for which permissions syncing request has been scheduled.",
 })
 
+var syncJobsRecordsTTLMinutes = env.Get("SYNC_JOBS_RECORDS_TTL_MINUTES", "15", "Duration to keep sync job completion records")
```
Comment (thenamankumar) [inline enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:47]: This can be integrated with site config.

**Comment by thenamankumar** [comment]
I have reviewed all the PRs. Will pushing the Provider Status to Redis be a very high load task that we need to do it via a separate routine? Can't it be done at each point where we are recording the provider status in first place? We do push many items to prometheus during the process and pushing things to redis should be quick. 

I am worried about the many places and functions this touches as it is passed up to so many levels making it a bit cumbersome to refactor later on.

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer.go (hunk 1 above)
Comment (bobheadxi) [reply enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:47]: It adds a bit of complexity - we must implement live-updating of the TTL of the log records, but I'll add this

**Comment by bobheadxi** [comment]
> Will pushing the Provider Status to Redis be a very high load task that we need to do it via a separate routine? Can't it be done at each point where we are recording the provider status in first place? 

I may be reading this wrong, but [`sum(increase(src_repoupdater_perms_syncer_success_syncs{type="user"}[1m]))` on s2](https://sourcegraph.sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%22now-6h%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum(increase(src_repoupdater_perms_syncer_success_syncs%7Btype%3D%5C%22user%5C%22%7D%5B1m%5D))%22,%22datasource%22:%22Prometheus%22,%22exemplar%22:true,%22requestId%22:%22Q-ba162a8e-9b6e-4d0f-aec0-c4a3b567ab54-0A%22%7D%5D) indicates we have 20-30 syncs per minute for user-centric perms sync (this metric doesn't seem to record repo-centric perms sync, which would account for more), so I thought it might be safer to do this async - that said, it is similar to the load we place on the DB so maybe a more frequent write at this rate could be fine, got a bit ahead of myself here 😛 

> We do push many items to prometheus during the process

note that generating prometheus metrics does _not_ push items to prometheus - prometheus will scrape the service for aggregated metrics on a regular time interval, every 30 seconds - and is not a comparable process to making a network request to set a record in Redis (though it is comparable to a DB write as I mentioned above)

**Review by unknwon — APPROVED** [review]

File: enterprise/cmd/repo-updater/internal/authz/sync_logs.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+			r.cache = rcache.NewWithTTL(syncJobsRecordsPrefix, ttlSeconds)
+			r.logger.Info("enabled records store cache", log.Int("ttlSeconds", ttlSeconds))
+		} else {
+			r.cache = noopCache{}
+			r.logger.Info("disabled records store cache")
+		}
+	})
+}
+
+// Add queues a record for the completion of a sync job.
```
Comment (kopancek) [inline enterprise/cmd/repo-updater/internal/authz/sync_logs.go:72]: I don't really understand this comment, can you rephrase? The method is called `Record` and here it says `Add queues a record` :) Should it say `Record queues a record` instead?

File: enterprise/cmd/repo-updater/internal/authz/sync_logs.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
+package authz
```
Comment (kopancek) [inline enterprise/cmd/repo-updater/internal/authz/sync_logs.go:1]: This file looks like new code, so I would like to see some unit tests, at least for the happy path.

File: enterprise/cmd/repo-updater/internal/authz/sync_logs.go (hunk 2 above)
Comment (bobheadxi) [reply enterprise/cmd/repo-updater/internal/authz/sync_logs.go:72]: Ah outdated comment - fixed!

**Review by pjlast — APPROVED** [review]

**Comment by mrnugget** [comment]
This feels like a temporary solution to get a quick win, IMHO. Why not build the proper thing? Can't we migrate the sync jobs to be actual database-backed worker jobs? That would be the ideal outcome. And if that's our goal then this seems like we make it harder for ourselves in the future to migrate to the proper solution.

File: enterprise/internal/authz/syncjobs/records_store.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	}
+}
+
+func (r *RecordsStore) Watch(c conftypes.WatchableSiteConfig) {
+	c.Watch(func() {
+		ttlMinutes := c.SiteConfig().AuthzSyncJobsRecordsTTL
+		if ttlMinutes == 0 {
+			ttlMinutes = defaultSyncJobsRecordsTTLMinutes
+		}
+		if !r.cacheTTL.CompareAndSwap(r.cacheTTL.Load(), int32(ttlMinutes)) {
```
Comment (mrnugget) [inline enterprise/internal/authz/syncjobs/records_store.go:51]: (These seem to be theonly two usees of `r.cacheTTL` -- looks like we don't need it)

Comment (bobheadxi) [reply enterprise/internal/authz/syncjobs/records_store.go:51]: I thought to do this in a 2-step process, because `Watch` procs on _any_ config change even if nothing relevant to the records store changed - this lets us check if a change is needed before acquiring a lock that blocks further writes. On second thought though it does seem unnecessary, will remove

**Comment by bobheadxi** [comment]
> This feels like a temporary solution to get a quick win, IMHO. Why not build the proper thing? Can't we migrate the sync jobs to be actual database-backed worker jobs? 

Unfortunately, I don't have the capacity to drive such an extensive migration yet, but we do need an endpoint available as part of https://github.com/sourcegraph/customer/issues/1534

Additionally, I'm not very convinced a database-backed job framework is a good choice here. Perms sync seems designed to queue and run ephemerally, such that if you just run it enough times it'll eventually get you to the right state, unless there's a major issue, in which case those will be surfaced in the last X jobs. There appears to be no real need to go back in time, nor for the jobs to be resilient because they get queued all the time and the point in time in which they get queued isn't very important.

**Comment by mrnugget** [comment]
> Additionally, I'm not very convinced a database-backed job framework is a good choice here. Perms sync seems designed to queue and run ephemerally, such that if you just run it enough times it'll eventually get you to the right state, unless there's a major issue, in which case those will be surfaced in the last X jobs. There appears to be no real need to go back in time, nor for the jobs to be resilient because they get queued all the time and the point in time in which they get queued isn't very important as long as they run eventually

Disagree. We want to display in the UI: last sync at, next sync at. We want to enqueue a next sync. We want to see the errors of the last sync. We want to have different priorities for syncs. All of that can be done with a database and a database-backed worker.

Querying the list of last sync jobs is another argument for that, IMHO.

**Comment by bobheadxi** [comment]
True, some of those components are in place already - for example we track the outcome of the last sync for a particular entity already - but next sync, priority, state of the queue etc isn't transparent at all.

Only discussion I've found is https://github.com/sourcegraph/sourcegraph/issues/16197 - I'll open up an issue for IAM

**Comment by bobheadxi** [comment]
See https://github.com/sourcegraph/sourcegraph/issues/44633
